docs/otpasswd.1
.\"
.\"   otpasswd(1) - One-Time Password Authentication System
.\"
.\"   Copyright (c) 2010-2013 Tomasz bla Fortuna
.\"
.\"   This file is part of OTPasswd.
.\"
.\"   OTPasswd is free software: you can redistribute it and/or modify
.\"   it under the terms of the GNU General Public License as published
.\"   by the Free Software Foundation, either version 3 of the License,
.\"   or any later version.
.\"
.\"   OTPasswd is distributed in the hope that it will be useful, but
.\"   WITHOUT ANY WARRANTY; without even the implied warranty of
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"   GNU General Public License for more details.
.\"
.\"   You should have received a copy of the GNU General Public License
.\"   along with OTPasswd.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\"   Author:  R Hannes Beinert & Tomasz bla Fortuna
.\"   Update:  02-Nov-13
.\"
.TH OTPASSWD 1 2013-11-02 "OTPasswd v0.8" "OTPasswd User Manual"
.\"

.SH NAME
otpasswd \- One-time password management utility
.\"

.SH SYNOPSIS
.B otpasswd
[\fIoptions\fR]
.\"

.SH DESCRIPTION
The \fIotpasswd\fR utility serves as the primary management tool
for the \fBOTPasswd\fR one-time password authentication package
for both users and administrators.
It allows the user to generate a key, print passcards,
and manage various configuration options.
.PP
In general, once \fBOTPasswd\fR has been installed,
the system administrator must determine where the system will keep
the user configuration data (user state).
The administrator will likely also configure the system by making various
policy choices,
such as how long passcodes must be, how complicated the static password should be,
which passcode alphabet will be used, etc.
Subsequently, users will probably also make some configuration choices
\- to the degree they are permitted by policy \-
and then they will generate a key and print some passcards.
All of these activities involve the use of the \fBotpasswd\fR(1) utility.
Once this initial configuration is complete,
the \fBOTPasswd\fR system will operate virtually autonomously.
Each time a user attempts to login to the system,
or otherwise authenticate with some service which uses
the \fBpam_otpasswd\fR(8) module,
the previously configured data is consulted,
a one-time passcode challenge will be issued,
and the user will have a chance to authenticate.
.\"
.SH OPTIONS
.SS User Administration
.TP
\fB\-k\fR, \fB\-\-key\fR
Generate a new key.
May be combined with the \fB\-\-config\fR (\fB\-c\fR) option.
If the user already has state information present,
\fBotpasswd\fR can either use the system default user state configuration,
or it can continue to use the the previously configured user state.
In this situation, the user is asked to express a preference.
Regardless, the user configuration may subsequently be modified through the
use of option flags to the \fBotpasswd\fR utility.
Once key generation is complete and the user state has been prepared,
\fBotpasswd\fR prints the first passcard and waits until the user
has made a note of the first few passcodes.
.\"
.TP
\fB\-r\fR, \fB\-\-remove\fR
Remove user state.
This completely removes any existing user configuration.
Note that if system policy requires \fBOTPasswd\fR use,
this action will effectively disable the user's login.
Furthermore, if policy also prohibits user key generation,
only the administrator will be able to restore access.
.\"
.SS Passcard Printing
.TP
\fB\-t\fR, \fB\-\-text\fR ( \fIcard\fR | \fIcode\fR )
Generate either one ASCII passcard, or a single passcode.
.\"
.TP
\fB\-l\fR, \fB\-\-latex\fR ( \fIcard\fR | \fIcode\fR )
Generate six (6) LaTeX passcards beginning with the specified passcard.
.\"
.SS Miscellaneous Passcode Operations
.TP
\fB\-P\fR, \fB\-\-prompt\fR ( \fIcard\fR | \fIcode\fR )
Display authentication prompt for the specified passcode.
.\"
.TP
\fB\-a\fR, \fB\-\-authenticate\fR \fIpasscode\fR
Attempt authentication with the actual passcode specified.
This might be used for testing or scripted authentication.
.\"
.TP
\fB\-s\fR, \fB\-\-skip\fR ( \fIcard\fR | \fIcode\fR )
Skip forward to a specific point in the passcode sequence.
The next authentication will request either the specified passcode,
or the first passcode on the specified passcard.
.\"
.TP
\fB\-w\fR, \fB\-\-warning\fR
Display any current warnings.
For example, if the user is currently on the last printed passcard.
It is useful to include this option in the user shell start-up commands
to stay informed of any \fBOTPasswd\fR system conditions which may
affect the user's ability to login.
.\"
.SS Configuration
.TP
\fB\-i\fR, \fB\-\-info\fR
Display current user state configuration.
.\"
.TP
\fB\-\-info-key\fR
Display private sequence key and passcode counter.
.\"
.TP
\fB\-c\fR, \fB\-\-config\fR \fIitem\fR
Set various flags which regulate \fBOTPasswd\fR system behavior:
.RS
.RS +4m
.TP +8m
\fBshow=\fR( \fIon\fR | \fIoff\fR )
Configure passcode visibility during authentication.
.TP +8m
\fBalphabet=\fR( \fIID\fR | \fIlist\fR )
Sets the \fIID\fR of the alphabet to be used.
The \fIlist\fR keyword may be used to display
a list of all available alphabets.
By default, a 64-character alphabet is used.
.TP +8m
\fBcodelength=\fR\fIlength\fR
Set the number of passcode characters to \fIlength\fR.
\fIlength\fR may be any value between 2 and 16.
Four (4) character passcodes are used by default.
.TP +8m
\fBsalt=\fR( \fIon\fR | \fIoff\fR )
Meaningful only during key generation.
Enables or disables salting of the passcode counter.
Setting this option to \fIoff\fR will make \fBOTPasswd\fR
compatible with the PPPv3.1 specification.
This has the additional consequence of increasing the number
of available passcards while (theoretically) lowering the
security of the system.
.TP +8m
\fBdisable=\fR( \fIon\fR | \fIoff\fR )
Disable the user.
Unlike \fB--remove\fR, this will \fInot\fR delete the user's state information.
.TP +8m
\fBcontact=\fR\fIstring\fR
Set contact information to \fIstring\fR.
This is used when passcodes are sent via an out-of-band channel.
The exact interpretation of \fIstring\fR will depend on the user configuration,
but could be a phone number, IM username, email address, etc.
Use an empty string ("") to unset this value.
.TP +8m
\fBlabel=\fR\fItitle\fR
Set a title for the generated passcards.
This will be displayed at the top of every passcard.
Use an empty string ("") to clear this value,
in which case the current hostname will be used by default.
.RE
.RE
.\"
.TP
\fB\-p, \fB\-\-password[=\fR\fIpassword\fR]
Set the static password. Argument is optional,
if not given otpasswd will ask you for password in a secure manner.
Enter twice the empty password to delete it.
.\"
.TP
\fB\-u\fR, \fB\-\-user\fR ( \fIusername\fR | \fIUID\fR )
Specify the user account upon which to operate.
Normal users may only change their own account, while the superuser may
change any account with this option.
Either a symbolic \fIusername\fR or a numeric \fIUID\fR may be specified.
(Administrator only)
.\"
.SS General
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Display more information about program operation. 
Use twice for even more information.
.\"
.TP
\fB\-\-version\fR
Display license, warranty, version and author information.
.\"
.TP
\fB\-h\fR, \fB\-\-help\fR
Program usage and option help information.
.\"
.TP
\fB\-\-check\fR
Obsolete option, moved to the agent_otp executable.
.\"
.SS Parameter References
When an option accepts a passcode reference (\fIcode\fR),
it can be specified in any of the following ways:
.PD 0
.RS +4m
.TP +10m
#
Passcode number.
Each passcode has a unique decimal number,
starting with the first passcode generated.
.TP +10m
RRC\fB[\fR#\fB]\fR
.TP +10m
CRR\fB[\fR#\fB]\fR
Passcode address.
Each passcode has a unique address comprised of the
column \fIC\fR (a letter) and row \fIRR\fR (one or more digits) on a
specific passcard number (\fI#\fR).
For example, 'F3[45]' refers to the passcode at column \fIF\fR
and row \fI3\fR of passcard \fI45\fR.
.TP +10m
\fBcurrent\fR
The passcode which will be used for the next authentication.
.RE
.PD
.\"
Passcard references (\fIcard\fR) may be specified as follows:
.PD 0
.RS +4m
.TP +10m
\fB[\fR#\fB]\fR
Passcard number.
.TP +10m
\fB[current]\fR
The passcard containing the current passcode.
.TP +10m
\fB[next]\fR
The first unprinted passcard.
Brackets are optional.
.RE
.PD
.\"

.SH SECURITY NOTES
This section needs to be completed (FIXME).
Until this section is written, see the docs/security file which
is distributed with the \fBOTPasswd\fR sources.
.\"

.SH EXIT STATUS
\fBotpasswd\fR will return zero on success and non-zero on failure.
This can be used for scripting
(e.g. see the \fB\-\-authenticate\fR/\fB\-a\fR option).
.\"
.\"   RETURN VALUE       [Normally only in Sections 2, 3]
.\"   ERRORS             [Typically only in Sections 2, 3]
.\"   ENVIRONMENT
.\"

.SH FILES
.TP
/etc/otpasswd
The directory for \fBOTPasswd\fR configuration files.
If \fBOTPasswd\fR is configured to use a global flat-file
database, then this file resides here, also.
The directory should be owned by the special \fIotpasswd\fR
system user,
and should not be accessible for normal users (mode 0700).
.\"
.TP
/etc/otpasswd/otpasswd.conf
The main \fBOTPasswd\fR system configuration file.
This file contains key-value configuration parameters that
govern the operation of the system.
In particular, it contains the state database configuration,
system-wide defaults, PAM configuration and policy configuration.
When the backend user database is configured to use MySQL or LDAP,
this file will contain privileged information and
must be not be accessible for normal users.
.\"
.TP
/etc/otpasswd/otshadow
The system-wide user database, used only when the \fBotpasswd.conf\fR(5)
configuration file specifies the use of a global database backend.
The file contains state information for all users,
including keys, flags, etc, and must not be accessible
to normal users.
.\"
.TP
$HOME/.otpasswd
This file is only used when the system configuration file
\fBotpasswd.conf\fR(5) specifies that state information is
to be maintained in user home directories.
This has the same format as the \fBotshadow\fR(5) file above,
except it only contains information for a single user.
.\"
.TP
/etc/pam.d/otpasswd_login
The PAM prototype configuration for \fBOTPasswd\fR.
If this file is \fIinclude\fRd in a PAM configuration for any
service (like \fBSSH\fR), the \fBpam_otpasswd\fR(8) PAM module will be
used to ask the user for a passcode just after the normal \fBpam_unix\fR(8)
authentication mechanisms are applied.
.\"
.TP
/lib/security/pam_otpasswd.so
The \fBOTPasswd\fR system PAM module.
This module is dynamically loaded by \fBPAM\fR(7) when an \fBOTPasswd\fR
one-time password authentication has been configured.
Various runtime options exist for this module to modify
operation.  For more information, see \fBpam_otpasswd\fR(8).
.\"
.\"   VERSIONS           [Normally only in Sections 2, 3]
.\"

.SH COMPATIBILITY
The \fBOTPasswd\fR authentication system is compatible with the
"Perfect Paper Passwords" specification version 3 (PPPv3) as
developed by the Gibson Research Corporation.
See DOCUMENTATION, below.
.PP
Note that for \fBOTPasswd\fR to operate in a manner which is
strictly compatible with PPPv3, it is important to specify
the \fB\-\-config salt=no\fR option during key generation.
A compatible key will generate passcards and passcodes
which are interoperable with other PPPv3-compliant applications.
For a list of such applications, see
<http://www.grc.com/ppp/software.htm>.
.PP
\fBOTPasswd\fR does \fInot\fR support any earlier versions
of the PPP specification.
.\"
.\"   CONFORMING TO
.\"   NOTES
.\"   BUGS
.\"

.SH EXAMPLES
Every user must generate a key in order to use \fBOTPasswd\fR.
Depending on the value of the \fBSALT_DEF\fR parameter in the
\fBotpasswd.conf\fR(5) file, a cryptographic salt may be used
to generate the passcode counter by default.

ChangeLog
ChangeLog / TODO
================

[+] New functionality
[!] Fixed error
[%] Found bug
[*] Changes in existing functionality
[-] Feature in TODO
[?] New feature in TODO
[$] Idea which was given up

Trying to sort tasks according to their priority.

2013-10-23 0.8
	Preparing distributable, convoyable, usable and likable version.
	Big version jump because of quite a long time from previous changes.
	0.8 can be thought as a release candidate for 0.9 - I'm trying to keep
	versioning simple to ease packaging.

	* [!] In not mandatory mode will fail if user HOME is not found.
	* [%] PAM module maybe should mmap memory not to let key be additionaly swapped to disc.
	   It's is on the disc in plaintext but the less copies the better.

	* [*] By default running in DB=user. Trying to get it work out-of-the-box.
	* [*] Fixing default options to let otpasswd be more easily tested.
	* [*] Updated package name convention, new package script
	* [*] Move sources to src/ subdirectory.

	* [-] Approach to agent data passing is overcomplicated. stdin/stdout substitution
	      causes problems with communication and as it turns out - valgrind. Would better
	      keep stdin/stdout, but maybe not use them(?) Best to review.
        * [+] Review documentation and ensure it's up to date.
	* [+] Update Polish translation
	* [-] Relicense so it can be used more universally - will require changing of sha256
              implementation, but it would be best to use system openssl lib probably.
	* [+] Check how SUID version works.
	      Update: Was fine.
	* [-] Update defaults for OOB. Explain better how to configure OOB messaging.
	      Add otpasswd_oob checks to --check-config


2010-03-31 0.7_rc2
Done:
        * [!] On DB=user when administrator creates a key with -u option user
              can't access that state.
        * [!] Fixed security bug in static password handling.
        * [+] Dropping dependency on GMP. Cleaning up "num" interface.
          - Clean memory so there're no key/counter leaks
        * [+] Do not store things in mpz_t which don't need it (spass)
        * [+] Sanitize gettext environment
	* [+] Accept 2G[2] passcode specification entries.
	* [+] All previous functionality is now finally implemented.
	      All implemented testcases where successfully run after
	      The Split.
	* [+] Backward skipping removed from configuration.

        * [+] Printing with -l when skipped to the last passcard fixed .
        * [+] OOB Channel usage time updates + DoS security
        * [+] Ensure config privilege checking
        * [+] --info should print information whether static password is set.
        * [+] Write testcases for static password
        * [+] Make otpasswd work with su. Tested under Gentoo and works BUT
              without OOB (permission denied)
        * [+] "Problem solving section" in docs, mentioning use of -v on
              errors in the first place. This was done on the webpage.
        * [+] policy checking regarding salt. (won't fail sometimes)
              (Seems ok to me. Tested)

        * [$] Global DB should work with users which aren't in /etc/passwd.
              Should it?
        * [$] Add error_t and bool_t to differentiate differently
              used ints? (bool_t is very rarely used. It will better be left
              to documentation)
In progress:
        * [?] Cleanup PPP interface. This should look like follows.
              "State" is class implementing some basic features of
              state management. From this 'class' ppp is derived.
              ppp implements high-level functions which should be used
              explicitly to manage state information unless something
              more fine-grained is necessary. State on the other hand
              uses 'db' backend for operating on files/databases.
              * Lots of things done in direction to finish this task
        * [?] Security: Do we need to change something regarding ulimits?

TODO Major:
        * [?] Double passcode logging scheme
        * [%] Parameter -c alphabet=3,codelenght=8 is sometimes accepted
              but doesn't work.
        * [?] Parametrize PAM messages.
        * [?] Static password expire warnings + enforcement.
        * [?] Logging warnings printed to user at WARN level
        * [?] Key quality checking (duplicates)
        * [?] Check if lock files are links. if so. fail. Or rather always unlink
              before overwriting.
        * [?] Logging into syslog from utility if agent is SUID.
        * [?] Admin function to scan all states for things invalidating
              policy. This could be implemented in agent similar to
              --check-config
        * [-] Scan all FIXME/TODO entries (Especially OOB related)
        * [?] SELinux compatibility.
        * [?] Import/export of ascii state lines...
        * [?] Check if User can lock OTPasswd when working with su (stdout
        issue). We must NOT keep locks when outputing something to user.
        user
        * [?] OTPasswd when used with su should probably elevate it's rights
        with setuid() and then return back to original rights.
        * [?] I don't remember how I've done it and therefore I need to review
              security aspects of DB=user mode when user can mangle state file
              which is then accessed by uid0 PAM process.

Todo Minor:
        * [?] Will we be pppv3 compatible on big-endian?
        * [?] Sourced key generation (from file, string...)
        * [?] Improve LaTeX output (some colors, borders?)
        * [?] Incorporate SSH key fingerprints on passcards?
              Use some file as passcard background, allow hook
              regenerating this passcard
        * [-] Keep lines below 75 columns? Or 80?



2010-01-24 v0.5rc1
Done:
        * [+] Check bit distribution for alphabets not divisible by 2
        * [+] Remove dont-skip option.

utility/otpasswd.c
/**********************************************************************
 * otpasswd -- One-time password manager and PAM module.
 * Copyright (C) 2009-2013 by Tomasz bla Fortuna <bla@thera.be>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with otpasswd. If not, see <http://www.gnu.org/licenses/>.
 **********************************************************************/

#ifndef PROG_VERSION
#define PROG_VERSION "0.8"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#include <assert.h>
#include <unistd.h> /* chdir, environ */

/* OTP Agent */
#include "agent_interface.h"

/* Utility headers */
#include "print.h"
#include "nls.h"

/* Program functions / helpers */
#include "actions.h"

/* Constants used in PPP */
#include "ppp_common.h"

static const char *_program_name(const char *argv0)
{
	const char *pos;
	if (!argv0)
		return "otpasswd";

	pos = strrchr(argv0, '/');
	if (pos)
		return pos+1;
	else
		return argv0;
}

/** Print author, license and quit. */
static void _show_license(void)
{
	int cnt=0;

	printf(
		_("OTPasswd - One-Time Password Authentication System.\n"
		  "Version %s \n"
		  "Copyright (C) 2009, 2010 Tomasz bla Fortuna <bla@thera.be>\n"
		  "\n"
		  "This program is free software: you can redistribute it and/or modify\n"
		  "it under the terms of the GNU General Public License as published by\n"
		  "the Free Software Foundation, either version 3 of the License, or\n"
		  "(at your option) any later version.\n"
		  "\n"
		  "This program is distributed in the hope that it will be useful,\n"
		  "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
		  "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
		  "GNU General Public License for more details.\n"
		  "\n"
		  "You should have received a copy of the GNU General Public License\n"
		  "along with this program in a LICENSE file.\n"),
		  PROG_VERSION
	);
	printf("\n\n");

	printf(_("OTPasswd was build with following options enabled:\n"));
#if DEBUG
	printf(_("DEBUG "));
	cnt++;
#endif 

#if USE_NLS
	printf(_("NLS "));
	cnt++;
#endif 

#if PROFILE 
	printf(_("PROFILE "));
	cnt++;
#endif 

	if (!cnt) {
		printf(_("NONE"));
	}
	printf("\n");

}

/** Show program parameters */
static void _show_usage(int argc, const char **argv)
{
	const char *prog_name =	_program_name(argc >= 2 ? argv[0] : NULL);
	fprintf(stdout, _(
		"Usage: %s [options]\n"
		"Actions:\n"
		"  -k, --key\n"
		"           Generate a new key. You can pass\n"
		"           -c along and set length or alphabet\n"
		"  -r, --remove\n"
		"           Remove key, and disable OTP.\n"
		"\n"
		"  -s, --skip <which>\n"
		"           Skip to a card specified as an argument.\n"
		"  -t, --text <which>\n"
		"           Generate either one ascii passcard\n"
		"           or a single passcode depending on argument. \n"
		"  -l, --latex <which>\n"
		"           Generate a LaTeX output with 6 passcards\n"
		"           starting with the specified one\n"
		"  -a, --authenticate <passcode>\n"
		"           Try to authenticate with given passcode\n"
		"  -w, --warning\n"
		"           Display warnings (ex. user on last passcard)\n"
		"  -P, --prompt <which>\n"
		"           Display authentication prompt for given passcode\n"
		"\n"
		"Where <which> might be one of:\n"
		"  number         - a decimal number of a passcode\n"
		"  [number]       - a passcard number\n"
		"\n"
		"  CRR[number] or RRC[number] \n"
		"                 - a passcode in passcard of a given number.\n"
		"                   C is column (A through G), RR - row (1..10)\n"
		"  current        - passcode used for next time authentication\n"
		"  [current]      - passcard containing current passcode\n"
		"  next or [next] - first, not yet printed, passcard\n"
		"\n"
		"Configuration:\n"
	        "  -i, --info\n"
	        "           Print current configuration and state.\n"
	        "      --info-key\n"
	        "           Print key and counter used for generating passcodes.\n"
		"           Warning: This will print private data.\n"
		"  -c, --config <arg>\n"
		"           Can be passed multiple times. Manages various settings:\n"
		"           show=<on|off> configure visibility of passcode during\n"
		"                         authentication.\n"
		"           alphabet=<ID|list>\n"
		"                         select passcode alphabet. Use 'list' argument\n"
		"                         to get IDs of available alphabets.\n"
		"           codelength=<length>\n"
		"                         select passcode length.\n"
		"\n"
		"           contact=X     Set contact info (e.g. phone number) with\n"
		"                         which to receive a passcode during authentication.\n"
		"                         Details depends configuration. Use \"\" to disable.\n"

		"           label=X       Set a caption to use on generated passcards.\n"
	        "                         Use \"\" to set default (hostname)\n"
		"\n"
		"           salt=<on|off>\n"
		"                         Meaningful only during key generation.\n"
		"                         Enable/disable salting of a passcode counter.\n"
		"                         Disabling this option will make program\n"
		"                         compatible with PPPv3.1 and will increase\n"
		"                         available passcard number at the cost of\n"
		"                         (theoretically) less security.\n"
		"\n"
		"           disable=<on|off>\n"
		"                         Disable user without removing his data.\n"
		"\n"
		"  -p, --password\n"
		"           Set static password. (optionally: --password=<pass>)\n"
		"  -u, --user <username|UID>\n"
		"           Operate on state of specified user. Administrator-only option.\n"
		"  -v, --verbose\n"
		"           Display more information about what is happening.\n"
		"           Try double -v -v for even more information.\n"
		"  --version\n"
		"           Display license, warranty, version and author information.\n"
		"  -h, --help\n"
		"           Display this message\n"
		"  OBSOLETE:\n"
		"  --check  Run all testcases. (Moved to agent_otp executable)\n"

		"\n"
		"Notes:\n"
		"  Both --text and --latex can get \"next\" as a parameter which\n"
		"  will print the first not-yet printed passcard. Usage of this argument\n"
		"  is recommended as it enables OTPasswd to display warnings when user\n"
		"  reaches his last printed passcard.\n"
		"\nExamples:\n"
		"%s --config salt=off --key   generate new (not salted) key\n"
		"%s --text next               print first not-yet-printed passcard\n"
		"%s --text '[3]'              print third passcard to standard output\n"
		"%s --text current            print current passcode\n"
		"%s --config codelength=5     use 5-character long passcodes\n"
		"Generate a 6 passcards on A4 page using LaTeX:\n"
		"%s --latex next > tmp.latex\n"
		"pdflatex tmp.latex\n"),
	        prog_name, prog_name, prog_name, prog_name, prog_name, prog_name, prog_name
	);
}

/* Parsing of a flag argument is done here */
static int parse_flag(options_t *options, const char *arg)
{
	assert(arg != NULL);
	assert(options != NULL);

	/*** Booleans/specials support ***/
	if (strcmp(arg, "show=on") == 0)
		options->flag_set_mask |= FLAG_SHOW;
	else if (strcmp(arg, "show=off") == 0)
		options->flag_clear_mask |= FLAG_SHOW;
	else if (strcmp(arg, "salt=on") == 0)
		options->flag_set_mask |= FLAG_SALTED;
	else if (strcmp(arg, "salt=off") == 0)
		options->flag_clear_mask |= FLAG_SALTED;
	else if (strcmp(arg, "disable=off") == 0)
		options->flag_clear_mask |= FLAG_DISABLED;
	else if (strcmp(arg, "disable=on") == 0)
		options->flag_set_mask |= FLAG_DISABLED;
	else if (strcmp(arg, "alphabet=list") == 0) {
		if (options->action != OPTION_CONFIG) {
			printf(_("Only one action can be specified on the command line\n"
			         "and you can't mix alphabet listing with other flags.\n"));
			return 1;
		}
		options->action = OPTION_ALPHABETS; /* List alphabets instead of changing flags */

		/*** Label and contact support */
	} else if (strncmp(arg, "contact=", 8) == 0) {
		const char *contact = arg + 8;

		if (options->contact) {
			printf(_("Contact already defined\n"));
			return 1;
		}

		/* Store */
		options->contact = strdup(contact);
	} else if (strncmp(arg, "label=", 6) == 0) {
		const char *label = arg + 6;

		if (options->label) {
			printf(_("Label already defined\n"));
			return 1;
		}

		/* Store */
		options->label = strdup(label);

		/*** Integer argument support */
	} else {
		int tmp;
		if (sscanf(arg, "codelength=%d", &tmp) == 1) {
			if (tmp == -1)    /* Also illegal, but we use */
				tmp = -2; /* -1 to mark it's not set */

			if (tmp < 2 || tmp > 16) {
				printf(_("Invalid code length. Valid range is from 2 to 16.\n"));
				return 1;
			}
			options->set_codelength = tmp;
			
		} else if (sscanf(arg, "alphabet=%d", &tmp) == 1) {
			if (tmp == -1)    /* Also illegal, but we use */
				tmp = -2; /* -1 to mark it's not set */

			if (tmp < 0 || tmp >= ppp_alphabet_count) {
				printf(_("Invalid alphabet ID. Valid IDs are between 0 and %d.\n"), ppp_alphabet_count);
				return 1;
			}
			options->set_alphabet = tmp;
		} else {
			/* Illegal flag */
			printf(_("No such flag or illegal option (%s).\n"), arg);
			return 1;
		}
	}

	/* Verify user don't want to unset and set at the same time */
	if (options->flag_set_mask & options->flag_clear_mask) {
		printf(_("Illegal configuration defined.\n"));
		return 1;
	}


	return 0;
}

/* Parse command line. Ensure we do not put any wrong data into options,
 * that is - longer than expected or containing any illegal characters */
int process_cmd_line(int argc, char **argv, options_t *options)
{
	static struct option long_options[] = {
		/* Action selection */
		{"key",			no_argument,		0, OPTION_KEY},
		{"remove",		no_argument,		0, OPTION_REMOVE},
		{"skip",		required_argument,	0, OPTION_SKIP},
		{"text",		required_argument,	0, OPTION_TEXT},
		{"latex",		required_argument,	0, OPTION_LATEX},
		{"prompt",		required_argument,	0, OPTION_PROMPT},
		{"authenticate",	required_argument,	0, OPTION_AUTH},
		{"warning",		no_argument,		0, OPTION_WARN},

		/* Flags */
		{"info",		no_argument,		0, OPTION_INFO},
		{"info-key",		no_argument,		0, OPTION_INFO_KEY},
		{"config",		required_argument,	0, OPTION_CONFIG},
		{"password",		optional_argument,	0, OPTION_SPASS},
		{"user",		required_argument,	0, OPTION_USER},
		{"verbose",		no_argument,		0, OPTION_VERBOSE},
		{"check",		no_argument,		0, OPTION_CHECK},
		{"version",		no_argument,		0, OPTION_VERSION},
		{"help",		no_argument,		0, OPTION_HELP},

		{0, 0, 0, 0}
	};

	assert(argv != NULL);
	assert(options != NULL);

	while (1) {
		int option_index = 0;

		int c = getopt_long(argc, argv, "krs:t:l:P:a:wic:p::vu:h", long_options, &option_index);

		/* Detect the end of the options. */
		if (c == -1) {
			if (optind < argc) {
				printf(_("Garbage on command line. (%s)\n"), argv[optind]);
				goto error;
			}

			break;
		}

		/* Assert maximal length of parameter */
		if (optarg && (strlen(optarg) > 100)) {
			printf(_("Option argument too long!\n"));
			goto error;
		}

		switch (c) {
			/* Argument-less actions */
		case OPTION_VERSION:
			_show_license();
			goto error;

		case OPTION_WARN:
		case OPTION_KEY:
		case OPTION_REMOVE:
		case OPTION_CHECK:
		case OPTION_HELP:
		case OPTION_INFO:
		case OPTION_INFO_KEY:
		case OPTION_SPASS:
			/* Error unless there was flag defined for key generation */
			if (options->action != 0 &&
			    !(options->action == OPTION_CONFIG && c == OPTION_KEY)) {
				printf(_("Only one action can be specified on the command line.\n"));
				goto error;
			}

			if (optarg) {
				options->spass = strdup(optarg);
			} else {
				assert(!options->spass);
				options->spass = NULL;
			}
			options->action = c;
			break;

			/* Actions with argument */
		case OPTION_SKIP:
		case OPTION_TEXT:
		case OPTION_LATEX:
		case OPTION_PROMPT:
		case OPTION_AUTH:
			if (options->action != 0) {
				printf(_("Only one action can be specified on the command line.\n"));
				goto error;
			}
			options->action = c;

			if (optarg)
				options->action_arg = strdup(optarg);
			else
				options->action_arg = NULL;
			break;

			/* Actions with argument which can be connected with -k */
		case OPTION_CONFIG:
			if (options->action != 0 && 
			    options->action != OPTION_CONFIG &&
			    options->action != OPTION_KEY) {
				printf(_("Only one action can be specified on the command line.\n"));
				goto error;
			}

			if (options->action == 0)
				options->action = OPTION_CONFIG;

			assert(optarg != NULL);

			if (parse_flag(options, optarg) != 0)
				goto error;

			break;

		case '?':
			/* getopt_long already printed an error message. */
			_show_usage(argc, (const char **)argv);
			goto error;

		case OPTION_USER:
			assert(optarg != NULL);

			if (getuid() != 0) {
				printf(_("Only root can use the '--user' option\n"));
				goto error;
			}

			if (options->username) {
				printf(_("Multiple '--user' options passed\n"));
				goto error;
			}

			options->username = strdup(optarg);
			if (!options->username) {
				printf(_("Error (not enough memory) while copying username.\n"));
				goto error;
			}
			break;

		case OPTION_VERBOSE:
			options->verbose++;
			break;

		default:
			printf(_("Program error. You shouldn't end up here.\n"));
			assert(0);
			goto error;
		}
	}

	/* Check additional correctness */
	if (((options->flag_set_mask | options->flag_clear_mask) & FLAG_SALTED)
	    && (options->action != 'k')) {
		printf(_("The \"salt\" flag can only be specified during key creation!\n"));
		goto error;
	}

	return 0;

error:
	free(options->action_arg), options->action_arg = NULL;
	free(options->label), options->label = NULL;
	free(options->contact), options->contact = NULL;
	free(options->spass), options->spass = NULL;
	free(options->username), options->username = NULL;
	return 1;
}


static int perform_action(int argc, char **argv, options_t *options)
{
	agent *a = NULL;
	int retval = 1;

	/* Reconfigure printing subsystem; -v might be passed */
	switch (options->verbose) {
	case 0: 
		print_config(PRINT_STDOUT | PRINT_ERROR); 
		break;
	case 1: 
		print_config(PRINT_STDOUT | PRINT_WARN); 
		break; 

	default:
	case 2: 
		print_config(PRINT_STDOUT | PRINT_NOTICE); 
		break; 
	}

	/* Perform pre-action preparations (set user, check state existance) */
	if (options->action != 0 && options->action != OPTION_HELP) {
		retval = action_init(options, &a);
		if (retval != 0) {
			goto cleanup;
		}
	}     

	/* Perform action */
	switch (options->action) {
	case 0:
		printf(_("No action specified. Try passing -k, -s, -t or -l\n\n"));
		_show_usage(argc, (const char **) argv);
		retval = 1;
		goto cleanup;

	case OPTION_HELP:
		_show_usage(argc, (const char **)argv);
		retval = 0;
		break;

	case OPTION_KEY:
		retval = action_key_generate(options, a);
		break;

	case OPTION_REMOVE:
		retval = action_key_remove(options, a);
		break;

	case OPTION_SPASS:
		retval = action_spass(options, a);
		break;

	case OPTION_INFO:
	case OPTION_INFO_KEY:
	case OPTION_ALPHABETS:
		retval = action_info(options, a);
		break;

	case OPTION_CONFIG:
		retval = action_config(options, a);
		break;

	case OPTION_AUTH:
		retval = action_authenticate(options, a);
		if (retval == 0)
			retval = 1;
		else
			retval = 0;
		break;

	case OPTION_SKIP:
		retval = action_skip(options, a);
		break;

	case OPTION_WARN:
		retval = action_warnings(options, a);
		break;

	case OPTION_TEXT:
	case OPTION_LATEX:
	case OPTION_PROMPT:
		retval = action_print(options, a);
		break;

	case OPTION_CHECK:
		printf(_("OBSOLETE: Most of the checks were moved into agent executable.\n"));
		retval = 1;
		goto cleanup;

	default:
		printf(_("Program error. You shouldn't end up here.\n"));
		assert(0);
		retval = 1;
		goto cleanup;
	}


cleanup:
	if (a)
		action_fini(a);

	free(options->action_arg), options->action_arg = NULL;
	free(options->label), options->label = NULL;
	free(options->contact), options->contact = NULL;
	free(options->spass), options->spass = NULL;
	free(options->username), options->username = NULL;
	return retval;
}


/** Starts command-line interface of the utility */
static inline int run_cli(int argc, char **argv)
{
	int ret;

	/* Options passed to utility with command line */
	options_t options = {
		.action = 0,
		.action_arg = NULL,
		.label = NULL,
		.contact = NULL,
		.spass = NULL,

		.username = NULL,
		.verbose = 0,

		.flag_set_mask = 0,
		.flag_clear_mask = 0,
		.set_codelength = -1,
		.set_alphabet = -1,

		.user_has_state = 0,
	};

	/* 0) Initialize locale */
	locale_init();


	/* 1) Parse command line, store data in options struct */
	ret = process_cmd_line(argc, argv, &options);
	if (ret != 0) {
		/* Error already printed */
		return ret;
	}

	/* 3) Execute actions according to options struct */
	ret = perform_action(argc, argv, &options);

	return ret;
}


int main(int argc, char **argv)
{
	/* Check if somebody by accident didn't make
	 * utility SUID-root: */
	int real_uid = getuid();
	int set_uid = geteuid();

	int real_gid = getgid();
	int set_gid = getegid();

	if (real_uid != set_uid || real_gid != set_gid) {
		printf(
			"*** INSTALLATION ERROR ***\n"
			"This executable shouldn't have SUID or SGID flag enabled!\n"
			"Only the agent executable (agent_otp) when DB=global option\n"
			"is selected in config file should be SUID-root\n"
			"Dying.\n");
		return 1;
	}


	/* Pre-init debugging, and go! */
	locale_init();
	print_init(PRINT_NOTICE | PRINT_STDOUT, NULL);
	return run_cli(argc, argv);
}
